sds/src/scanner/mod.rs

```diff
@@ -96,113 +97,87 @@ impl Scanner {
             Box<dyn Fn() -> Vec<regex_automata::meta::Cache> + Send + Sync>,
         > = self.cache_pool.get();
 
-        // all matches, after overlapping rules have been filtered
-        let mut rule_matches = vec![];
+        // All matches, after some (but not all) false-positives have been removed.
+        // This is a vec of vecs, where each inner vec is a set of matches for a single path.
+        let mut rule_matches_list = vec![];
+
+        let mut excluded_matches = AHashSet::new();
 
         self.scoped_ruleset.visit_string_rule_combinations(
             event,
             ScannerContentVisitor {
                 scanner: self,
                 caches,
-                rule_matches: &mut rule_matches,
+                rule_matches: &mut rule_matches_list,
+                excluded_matches: &mut excluded_matches,
             },
         );
-
-        // TODO: multipass V0 logic will go here
-        self.apply_match_actions(event, rule_matches)
-    }
-
-    /// Apply mutations from actions, and shift indices to match the mutated values.
-    /// This assumes the matches are grouped by path (order doesn't matter) then sorted by start index.
-    fn apply_match_actions<E: Event>(
-        &self,
-        event: &mut E,
-        rule_matches: Vec<InternalRuleMatch<E::Encoding>>,
-    ) -> Vec<RuleMatch> {
         let mut output_rule_matches = vec![];
 
-        let mut rule_matches_iter = rule_matches.into_iter().peekable();
+        for rule_matches in &mut rule_matches_list {
+            // All rule matches in each inner list are for a single path, so they can be processed independently.
 
-        while rule_matches_iter.peek().is_some() {
-            let mut rule_match = rule_matches_iter.next().unwrap();
+            // TODO: remove the clone here
+            let path = rule_matches.first().unwrap().path.clone();
```

Comment:
This the path is extracted from the first match, won't it be better to remove it from the `InternalRuleMatch` and store it as a key of the `rule_matches_list`?

You won't need the clone anymore

Reply:
Good suggestion, changed in https://github.com/DataDog/dd-sensitive-data-scanner/pull/10/commits/2534554df734ef16e737b27c9b8fb080f668a8d4

Note that while I added another `clone`, it replaces an `.into_static()` which was basically a clone.